api/agent/agent.go

```diff
@@ -112,72 +113,42 @@ type agent struct {
 	shutonce sync.Once
 	shutdown chan struct{}
 
-	freezeIdleMsecs time.Duration
-	ejectIdleMsecs  time.Duration
-
 	// Prometheus HTTP handler
 	promHandler http.Handler
 }
 
 func New(da DataAccess) Agent {
-	// TODO: Create drivers.New(runnerConfig)
-	driver := docker.NewDocker(drivers.Config{
-		ServerVersion: "17.06.0-ce",
-	})
-
-	freezeIdleMsecs, err := getEnvMsecs("FN_FREEZE_IDLE_MSECS", 50*time.Millisecond)
-	if err != nil {
-		logrus.WithError(err).Fatal("error initializing freeze idle delay")
-	}
 
-	ejectIdleMsecs, err := getEnvMsecs("FN_EJECT_IDLE_MSECS", 1000*time.Millisecond)
+	cfg, err := NewAgentConfig()
 	if err != nil {
-		logrus.WithError(err).Fatal("error initializing eject idle delay")
-	}
-	if ejectIdleMsecs == time.Duration(0) {
-		logrus.Fatal("eject idle delay cannot be zero")
+		logrus.WithField("cfg", cfg).WithError(err).Fatal("error in agent config")
 	}
+	logrus.WithField("cfg", cfg).Info("agent starting")
 
-	logrus.WithFields(logrus.Fields{"eject_msec": ejectIdleMsecs, "free_msec": freezeIdleMsecs}).Info("agent starting")
+	// TODO: Create drivers.New(runnerConfig)
+	driver := docker.NewDocker(drivers.Config{
+		ServerVersion: cfg.MinDockerVersion,
+	})
 
 	a := &agent{
-		da:              da,
-		driver:          driver,
-		slotMgr:         NewSlotQueueMgr(),
-		resources:       NewResourceTracker(),
-		shutdown:        make(chan struct{}),
-		freezeIdleMsecs: freezeIdleMsecs,
-		ejectIdleMsecs:  ejectIdleMsecs,
-		promHandler:     promhttp.Handler(),
+		cfg:         *cfg,
+		da:          da,
+		driver:      driver,
+		slotMgr:     NewSlotQueueMgr(),
+		resources:   NewResourceTracker(),
+		shutdown:    make(chan struct{}),
+		promHandler: promhttp.Handler(),
 	}
 
+	protocol.BufPoolChunkSize = int64(cfg.MaxResponseSize)
```

Comment:
this one kinda smells. I see a lot of ways to skin the cat on this one. was this preferable to wrapping the reader or writer being passed off to dispatch [as stdin/stdout] in the same manner we're doing for stderr (`func_logger.go`)?

Reply:
yeah, I'll take a look at this one more time. I kind of started with global var like this because this was tied to system wide env var, but now I moved it under the agent, who sets it, so it surely looks a bit unsavory.